### ISSUE TITLE
ActionMailer `*_deliver` callbacks invoked regardless if mail sent

### ISSUE THREAD

**Issue by dan-jensen** [rails/rails#53155]
### Steps to reproduce

```ruby
class ExampleMailer < ActionMailer::Base

  before_deliver :handle_before_deliver
  after_deliver :handle_after_deliver

  def send_example
    return # assume this message has been re-scheduled, is no longer necessary, etc.

    mail()
  end

  def handle_before_deliver
    if message.kind_of?(ActionMailer::Base::NullMail)
      puts "UNEXPECTED: before_deliver was invoked even though nothing was sent or delivered"
    end
  end

  def handle_after_deliver
    if message.kind_of?(ActionMailer::Base::NullMail)
      puts "UNEXPECTED: after_deliver was invoked even though nothing was sent or delivered"
    end
  end

end

ExampleMailer.send_example.deliver_now

UNEXPECTED: before_deliver was invoked even though nothing was sent or delivered
UNEXPECTED: after_deliver was invoked even though nothing was sent or delivered
```

### Expected behavior

ActionMailer `*_deliver` callbacks are invoked only if mail was sent.

### Actual behavior

ActionMailer `*_deliver` callbacks are invoked regardless if mail was sent.

### System configuration
**Rails version**: 7.1.3.4

**Ruby version**: 3.3.1

### Feature details

The `*_delivered` callbacks were introduced in PR #47630. There was [some concern](https://github.com/rails/rails/pull/47630#issuecomment-1489517789) the names imply SMTP delivering (which is outside the scope of the Mail gem). This issue goes further by expressing concern the names at least imply SMTP sending (which is the core purpose of the Mail gem).

**Comment by dan-jensen** [comment]
@bensheldon love this feature, thanks for creating it. I just found this particular behavior unexpected. Do you agree?

I'm thinking we should add 3 guard clauses in [ActionMailer::Callbacks](https://github.com/rails/rails/blob/main/actionmailer/lib/action_mailer/callbacks.rb) that are similar to [the one in ActionMailer::Base](https://github.com/rails/rails/blob/main/actionmailer/lib/action_mailer/base.rb#L658): `return unless @_mail_was_called`. What do you think?

**Comment by bensheldon** [comment]
I think that makes sense. From reading through the code I think `@_mail_was_called` is doing double duty to determine if `mail` has been called _and_ to prevent `mail` from being called more than once. 

I wonder if the logic to determine if the callback should be triggered should specifically test for `ActionMailer::Base::NullMail`.

**Comment by dan-jensen** [comment]
@bensheldon yep, that would definitely work. But I was thinking we should use the same mechanism as ActionMailer::Base if possible. Could you help me understand why you suggest checking for `NullMail` instead? Are you concerned about relying on ActionMailer internals?

**Comment by bensheldon** [comment]
Maybe I misunderstood your comment. I took your comment to imply that `@_mail_was_called` should be used; I don't think that is the appropriate ivar because of its purpose. No objection to using ivars, I don't think that's the right ivar from my interpretation of the code's intent.

**Comment by dan-jensen** [comment]
@bensheldon actually I think we already agree. I'll explain.

Yes, I think the guard clause in the callbacks should be `return unless @_mail_was_called`. That seems the right ivar to me because I believe it dictates whether `message` is an instance of `ActionMailer::Base::NullMail`. I'm referring to [line 658](https://github.com/rails/rails/blob/main/actionmailer/lib/action_mailer/base.rb#L658):

`@_message = NullMail.new unless @_mail_was_called`

If that's correct, then we are talking about testing for the same thing. It's just a question of whether we want to rely on the ivar or the message class:

- ivar option: `return unless @_mail_was_called`
- message class option: `return if message.kind_of?(ActionMailer::Base::NullMail)`

Personally, I'm inclined to use the ivar option to match the other ActionMailer code, including line 658.

Am I totally missing something, or does this make sense?

**Comment by bensheldon** [comment]
@dan-jensen I think it would be best for you to open the PR, because I'm not really following the implementation you're proposing. Seeing the code would be much more helpful 💛 

Please ping me on the PR, though fyi I don't have a role in Rails so my opinions don't have any authority behind them 😁 I do think your bug report is valid.

fyi, though I don't want to propose jamming up your PR with extra scope, but for context about the bigger picture, there is this: #50858

(in case that would make it easier/harder to prevent the `after_deliver` callback)

**Comment by dan-jensen** [comment]
@bensheldon ah thanks, I wasn't aware of that PR. Seems like that should be merged first, because this change would touch the same code.

And yes, this is tricky because it seems the conditional we're talking about needs to be injected into `filters` when `set_callback` is called.

Maybe we should start with a quick note in the docs while we're waiting on #50858. Something like "*_deliver callbacks run regardless if `mail` is invoked". What do you think?

**Comment by bensheldon** [comment]
I don't have strong feelings either way. I found the PR where the ivar was introduced and I think you're correct on the purpose of that ivar: https://github.com/rails/rails/pull/8450

I think you should make the PR to fix it, and then we can talk about that.
